lib/util.c

```diff
@@ -2187,9 +2194,332 @@ EXPORTED void xsyslog_fn(int priority, const char *description,
+ *                    "location", "%s", "Dubuque",
+ *                    "richter", "%5.1f", 7.9)
+ *
+ * will return the string
+ *
+ *     event=earthquake location=Dubuque richter="  7.9"
+ *
+ * char *s = xsyslog_ev_fmt(event_name,
+ *                          key1, fmt1, val1,
+ *                          key2, fmt2, val2,
+ *                          ...);
+ *
+ * The fmt_n arguments are SINGLE sprintf-style formats.  These are used to format the
+ * corresponding val_n arguments; the formatted values are then
+ * escaped with xsyslog_ev_escape_value. (This is why the richter value in the example is quoted.)
+ *
+ * The "event_name" is similarly formatted as if it had been part of a triple:
+ *    "event", "%s", event_name
+ *
+ * A sequence of chunks is built, each with the form "key_n=escaped_formatted_val_n"
+ *
+ * The chunks are concatenated with single spaces in between.
+ *
+ * The resulting string is returned.  The CALLER is responsible for freeing it.
+ *
+ * WARNING:
+ *   the function only supports a subset of sprintf-style escapes.
+ *   in particular, each format string MUST contain EXACTLY ONE %-escape.
+ *   See xsyslog_ev_guess_printf_escape below.
+ *
+ * SPECIAL EXCEPTION:
+ *   if the argument list also contains .... "__sep__", "FOO" ....
+ *   the separator string "FOO" will be used in place of the single spaces.
+ *   For example:
+ *     xsyslog_ev_fmt("earthquake", "__sep__", "; ",
+ *                    "location", "%s", "Dubuque")
+ *
+ *   will return the string
+ *
+ *       event=earthquake; location=Dubuque
+ */
+EXPORTED char *xsyslog_ev_fmt(const char *event_name, ...)
+{
+    va_list ap;
+    va_start(ap, event_name);
+    char *s = xsyslog_ev_fmt_va(event_name, ap);
+    va_end(ap);
+    return s;
+}
+
+/* xsyslog_ev_fmt_va: va_list version of xsyslog_ev_fmt
+ *
+ * same as xsyslog_ev_fmt, except it gets an initialized va_list argument
+ * instead of a variable argument list
+ *
+ * The CALLER is responsible for freeing the returned string.
+ */
+static char *xsyslog_ev_fmt_va(const char *event_name, va_list ap)
+{
+    struct buf log = BUF_INITIALIZER;
+    struct buf val = BUF_INITIALIZER;
+    const char *sep = " ";  // string that separates kev=value pairs
+    const char *key;
+
+    // Add "event=(event-name)" to the log buffer
+    buf_setcstr(&val, event_name);
+    buf_printf(&log, "event=%s", xsyslog_ev_escape_value(&val));
+
+    /*
+     * Now handle the remaining arguments, which should come in threes:
+     *   key, format, value
+     *
+     * value will be buf_printf'ed with the format,
+     * and then the result will be escaped (if necessary)
+     *
+     * finally, the resulting string "key=escaped-formatted-value"
+     * will be appended to the log buffer
+     */
+    while ((key = va_arg(ap, const char *))) {
+        // special case: __sep__ means reconfigure the separator
+        if (!strcmp(key, "__sep__")) {
+            sep = va_arg(ap, const char *);
+        }
```

Comment:
Is this actually useful?  Surely if we're going to all this effort it's because we want our logs to be in a consistent format

Reply:
Probably not. I only threw it in as a possibly interesting suggestion.

YAGNI says to eliminate it and its associated tests.